django/contrib/postgres/search.py

```diff
@@ -219,6 +221,53 @@ def as_sql(self, compiler, connection, function=None, template=None):
         return sql, extra_params + params
 
 
+class SearchHeadline(Func):
+    function = 'ts_headline'
+    output_field = TextField()
+    options = None
+
+    def __init__(self, expression, query, **extra):
+        if not hasattr(query, 'resolve_expression'):
+            query = SearchQuery(query)
+        self.config = query.config
+        self.options = extra.get('options', self.options)
+        super().__init__(expression, query, **extra)
+
+    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
+        resolved = super().resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        if self.config:
+            if not hasattr(self.config, 'resolve_expression'):
+                resolved.config = Value(self.config).resolve_expression(query, allow_joins, reuse, summarize, for_save)
+            else:
+                resolved.config = self.config.resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        return resolved
+
+    def as_sql(self, compiler, connection, function=None, template=None, arg_joiner=None, **extra_context):
+        config_params, options_params = [], []
+        extra_context = {}
+        if template is None:
+            template = '%(function)s(%(config)s%(expressions)s%(options)s)'
+            extra_context.update({
+                'config': '',
+                'options': '',
+            })
+            if self.config:
+                config_sql, config_params = compiler.compile(self.config)
+                extra_context['config'] = '{}::regconfig, '.format(config_sql)
+            if self.options:
+                options = []
+                for option, value in self.options.items():
+                    pg_value = psycopg2.extensions.adapt(value).getquoted().decode()
```

Comment:
I might be misunderstanding something but the problem here is that the whole `options` string needs to be passed as a value to `execute`. Otherwise psycopg2 will apply invalid quoting, for example:

```python
 cursor.execute(".., '%s=%s, %s=%s'", ('StartSel', '<strong>', 'StopSel', '</strong>'))
```
*Note the required quoting around the options template string*

Would quote `StartSel` and `StopSel` which is invalid syntax: 
```
''StartSel'='<strong>', 'StopSel'='</strong>''
```

Using `psycopg2.extensions.adapt` allows us to apply the correct quoting for strings and proper conversion of booleans.

If we where to still wrap value with `Value` it would have to be used something like this:
```python
_, value_params = compiler.compile(value)
pg_value = psycopg2.extensions.adapt(value_params[0]).getquoted().decode()
 ```

And that feels kind of awkward.

Let me know if I'm missing something!

Reply:
@charettes Ideas welcome. Can you take a look? I really don't like the idea of using `psycopg2` in expressions.

Reply:
@felixxm, @hannseman is right we need to compile a single `text` value for all the options so I don't see a way around invoking `adapt` here like we do when _baking_ a `DEFAULT` string value

https://github.com/django/django/blob/22ce5d0031bd795ade081394043833e82046016c/django/db/backends/postgresql/schema.py#L34-L38